### ISSUE TITLE
The benchmarks are somewhat pointless

### ISSUE THREAD

**Issue by RealDolos** [jonschlinkert/repeat-string#7]
The benchmarks are somewhat pointless in so far as they only test the performance of the cache (and `.substr`). Because the cache will be generated once per fixture and after that it's merely cache performance being tested over and over again. So basically, only the very best case is ever tested.

Since the docs put so obvious emphasis on benchmarks, I thought it would be good to mention.

So here is a version I quickly made that outperforms your implementation somewhat in the each-call-is-different worst case and by quite a bit in the everything-is-cached best case you're currently testing [0]. BENCHMARKS!
It's also passing tests.

```js
var res, last;
function repeat /* aka native2 */ (str, num) {
  if (typeof str !== 'string') {
    throw new TypeError('expected a string');
  }
  if (num < 1) {
    return '';
  }
  if (num === 1) {
    return str;
  }

  var max = str.length * num;
  var same = last === str;
  if (same && res.length === max) {
    return res;
  }
  if (same && res.length > max) {
    return res.substr(0, max);
  }
  last = str;
  res = str.repeat(num);
  return res;
};
```

Now I won't claim it's better because the benchmarks are better. Micro-benchmarks are misleading at best, in particular because those strings never get actually used in any way (and therefore are not necessarily materialized/flattened ever either).
I will claim tho it's better because it's more straight forward and gives the VM a clearer opportunity to optimize `str.repeat()` regarding CPU and memory consumption.
Then again, just ***using `str.repeat()`*** without caring about hidden caches and such and which is blazing fast in itself is probably ***the right decision anyway in 99.999%*** of all use cases... premature optimization yadayada and such...

Since I mentioned tests...
There are no tests to explicitly test the cache is working as expected, btw, but that's another issue for another day.

You say in https://github.com/jonschlinkert/pad-left/issues/9, and I quote:

>I think this is terrible open source, code theft shouldn't be encouraged. The only reason I added benchmarks is to show that despite using code from a this library, left-pad and repeating still aren't as performant. This is what happens when you copy code you don't understand.

So, I will assume you won't be stealing my code I posted here, nor the basic concepts and ideas I used. Thanks for not stealing!

I also encourage you to file a police report against @sindresorhus for the theft of "teh codez!!1!" since only a total thieving :smile_cat:burglar would think it's OK to use MIT-licensed code in another MIT-licensed project! :policeman: 

[0]
In browser benchmarks:
https://jsperf.com/repeat-string-and-stuff
https://jsperf.com/repeat-string-and-stuffs-cache-perf

```text
$ node --version
v7.0.0

# benchmark/fixtures/2.js (43 bytes)
  repeat-string x 23,939,329 ops/sec ±0.86% (88 runs sampled)
  native2 x 28,613,449 ops/sec ±0.96% (90 runs sampled)

  fastest is native2 - 16.47% faster
  slowest is repeat-string - 14.14% slower

# benchmark/fixtures/3.js (43 bytes)
  repeat-string x 11,821,146 ops/sec ±2.59% (86 runs sampled)
  native2 x 27,296,621 ops/sec ±6.48% (82 runs sampled)

  fastest is native2 - 19.12% faster
  slowest is repeat-string - 16.05% slower

# benchmark/fixtures/10.js (46 bytes)
  repeat-string x 20,608,420 ops/sec ±0.83% (92 runs sampled)
  native2 x 27,625,402 ops/sec ±0.83% (89 runs sampled)

  fastest is native2 - 33.51% faster
  slowest is repeat-string - 25.10% slower

# benchmark/fixtures/50.js (46 bytes)
  repeat-string x 20,920,148 ops/sec ±1.23% (90 runs sampled)
  native2 x 28,225,675 ops/sec ±0.82% (89 runs sampled)

  fastest is native2 - 36.85% faster
  slowest is repeat-string - 26.93% slower

# benchmark/fixtures/250.js (47 bytes)
  repeat-string x 12,168,990 ops/sec ±4.88% (81 runs sampled)
  native2 x 27,933,133 ops/sec ±3.70% (82 runs sampled)

  fastest is native2 - 48.43% faster
  slowest is repeat-string - 32.63% slower

# benchmark/fixtures/2000.js (48 bytes)
  repeat-string x 19,984,142 ops/sec ±1.02% (91 runs sampled)
  native2 x 27,415,366 ops/sec ±1.05% (86 runs sampled)

  fastest is native2 - 35.18% faster
  slowest is repeat-string - 26.03% slower

# benchmark/fixtures/20000.js (35 bytes)
  repeat-string x 19,344,269 ops/sec ±1.06% (89 runs sampled)
  native2 x 29,343,416 ops/sec ±0.96% (86 runs sampled)

  fastest is native2 - 43.00% faster
  slowest is repeat-string - 30.07% slower
```

Patched benchmarked for percentages
```js
  suite.on('complete', function() {
    function getHz (bench) {
          return 1 / (bench[0].stats.mean + bench[0].stats.moe);
    }
    let fastest = this.filter('fastest');
    let slowest = this.filter('slowest');
    let diff = ((getHz(fastest) / getHz(slowest)) - 1) * 100;
    let rdiff = (1 - (getHz(slowest) / getHz(fastest))) * 100;
    console.log('  fastest is', colors.green(fastest.map("name")),
      `- ${diff.toFixed(2)}% faster`
    );
    console.log('  slowest is', colors.red(slowest.map("name")),
      `- ${rdiff.toFixed(2)}% slower`
    );
```

**Comment by jonschlinkert** [comment]
This issue was closed as it included language that was ~~clearly~~ perceived as an effort to troll. The offending section of the OP was deleted, and marked as `DELETED` but the rest of the comment retains intact.  

Unfortunately, since you also cross-posted a comment similar to the deleted text on another user's project - especially given that I blocked this particular user a long time ago, you have been blocked personally.  However, if this is just a misunderstanding, and you sincerely would like to contribute and collaborate on improving this library, you can find my email on my profile and contact me directly, I would be happy to discuss this with you privately, and of course apologize and unblock you and the issue.

Regarding the code suggestions themselves, I appreciate the attempt to find ways to optimize and even took the time to hand-edit the benchmarks to add "slowest time" (a nice touch I might add, although I'm curious why repeating was removed from the revised benchmarks. I'm assuming it's because repeating was irrelevant to what you were trying to determine), however:
1. there is zero drawback to caching the string (the thesis if your issue), which leaves only potential upside
2. your assuming that the function won't be called more than once in a process, I'm not sure where that idea comes from, but I can think of more use cases for calling the function more than once, than for calling just once (like by command line I guess?), which tells us one thing: if we can optimize around both cases with the least downside possible, why not just optimize and avoid trying to make assumptions that we can't possibly be correct about?
3. your code depends on `String.prototype.repeat()` existing, which requires a shim before ES6. 
4. when I ran the benchmarks with your code, the current code in repeat-string is just as performant, faster in some cases. That said, I'm not opposed to using `String.prototype.repeat()`, but I don't see the benefit of it yet. 

> Micro-benchmarks are misleading at best, in particular because those strings never get actually used in any way (and therefore are not necessarily materialized/flattened ever either) ... premature optimization yadayada and such...

These are just my opinions, but in a micro-library that does one specific thing, especially something as fundamental as repeating a string:
- Optimizations are never premature (and if they were, when would be the right time? Should we wait for 200m downloads? Unfortunately this premature-optimization concept is way overinflated and rarely applied correctly. )
- Micro-benchmarks (in a micro-lib) are really just benchmarks. I can't think of a better use case for benchmarks than a lib that does only one thing. The definition of low-hanging fruit for optimizations.
- Although the benchmarks were useful for making optimizations, they are useful for providing insight into how this library performs in comparison to other libraries that do the exact same thing. and again, since there are no moving parts or differing goals (it's just "repeating a string" n times) for any of the compared libraries (or native method), the benchmarks are as relevant as you can get.
